File: packages/a2a/src/task-translator.ts
/**
 * Lifecycle translation between PolyMesh and A2A task states
 * (PM-V6-SPEC §A.6.3–§A.6.5, §A.9.3, §A.10.6, §A.17.4, §A.18).
 */
import { createHash } from "node:crypto";
import { existsSync, readFileSync, writeFileSync } from "node:fs";

import { uuidv7 } from "@latticeag/polymesh-broker";

import { skillNameFromCapabilityName, fidelityClause } from "./card-mapper.js";
import { mapA2ATaskErrorToPolyMesh } from "./errors.js";
import type {
  A2AArtifact,
  A2ATask,
  A2ATaskState,
  PolyMeshTaskState,
} from "./types.js";
import { TERMINAL_A2A_STATES, TERMINAL_POLYMESH_STATES } from "./types.js";

/** §A.6.3 forward projection (PolyMesh → A2A). */
export const POLYMESH_TO_A2A_STATE: Readonly<Record<PolyMeshTaskState, A2ATaskState>> =
  Object.freeze({
    SUBMITTED: "submitted",
    REJECTED: "failed",
    ACCEPTED: "working",
    QUEUED: "working",
    RUNNING: "working",
    WAITING: "working",
    SUCCEEDED: "completed",
    FAILED: "failed",
    CANCELLED: "canceled",
  });

/**
 * §A.6.3 inverse used for outbound status consumption (§A.9.3).
 * `submitted` is treated as `ACCEPTED` once the send ack has been received.
 */
export const A2A_TO_POLYMESH_STATE: Readonly<Record<A2ATaskState, PolyMeshTaskState>> =
  Object.freeze({
    submitted: "ACCEPTED",
    working: "RUNNING",
    completed: "SUCCEEDED",
    failed: "FAILED",
    canceled: "CANCELLED",
  });

export function isTerminalA2AState(state: A2ATaskState): boolean {
  return TERMINAL_A2A_STATES.includes(state);
}

export function isTerminalPolyMeshState(state: PolyMeshTaskState): boolean {
  return TERMINAL_POLYMESH_STATES.includes(state);
}

/**
 * Monotonic ordering rank (§A.10.6). Terminal states share the top rank so a
 * stale poll can never move a task between two terminal outcomes.
 */
const STATE_RANK: Readonly<Record<PolyMeshTaskState, number>> = Object.freeze({
  SUBMITTED: 0,
  ACCEPTED: 1,
  QUEUED: 2,
  RUNNING: 3,
  WAITING: 3,
  REJECTED: 9,
  SUCCEEDED: 9,
  FAILED: 9,
  CANCELLED: 9,
});

export function polyMeshStateRank(state: PolyMeshTaskState): number {
  return STATE_RANK[state];
}

/**
 * Tracks last-applied state for one outbound task and rejects regressive or
 * post-terminal updates (§A.10.6).
 */
export class MonotonicStateTracker {
  private current: PolyMeshTaskState;

  constructor(initial: PolyMeshTaskState = "SUBMITTED") {
    this.current = initial;
  }

  get state(): PolyMeshTaskState {
    return this.current;
  }

  get terminal(): boolean {
    return isTerminalPolyMeshState(this.current);
  }

  /** Returns true when the update was applied; false when discarded as stale. */
  apply(next: PolyMeshTaskState): boolean {
    if (this.terminal) return false;
    if (polyMeshStateRank(next) < polyMeshStateRank(this.current)) return false;
    if (next === this.current) return false;
    this.current = next;
    return true;
  }
}

export interface TranslatedTaskEvent {
  state: PolyMeshTaskState;
  terminal: boolean;
  progress?: number;
  message?: string;
  result?: unknown;
  error?: ReturnType<typeof mapA2ATaskErrorToPolyMesh>;
}

/**
 * Translate one observed A2A task snapshot into a mesh-side lifecycle event
 * (§A.9.3). Artifact extraction prefers the first `data` part.
 */
export function translateTaskEvent(task: A2ATask): TranslatedTaskEvent {
  const state = A2A_TO_POLYMESH_STATE[task.status.state];
  const event: TranslatedTaskEvent = {
    state,
    terminal: isTerminalPolyMeshState(state),
  };
  if (typeof task.status.progress === "number") event.progress = task.status.progress;
  if (task.status.message) event.message = task.status.message;
  if (state === "SUCCEEDED") event.result = extractArtifactJson(task.artifacts);
  if (state === "FAILED") event.error = mapA2ATaskErrorToPolyMesh(task.status.error);
  return event;
}

/** `ExtractArtifactJson` (§A.9.3): first JSON data part, else first text part. */
export function extractArtifactJson(artifacts: readonly A2AArtifact[] | undefined): unknown {
  if (!artifacts || artifacts.length === 0) return undefined;
  for (const artifact of artifacts) {
    for (const part of artifact.parts ?? []) {
      if (part.type === "data" && part.data !== undefined) return part.data;
    }
  }
  for (const artifact of artifacts) {
    for (const part of artifact.parts ?? []) {
      if (part.type === "text" && part.text !== undefined) return part.text;
    }
  }
  return undefined;
}

// ---------------------------------------------------------------------------
// §A.6.5 / §A.17.4 shared task id space
// ---------------------------------------------------------------------------

/** UUIDv7 shape, case-insensitive (§A.6.5). */
export const UUIDV7_PATTERN =
  /^[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;

export function isUuidV7(value: string): boolean {
  return UUIDV7_PATTERN.test(value);
}

/** Durable `local_task_id ↔ remote_task_id` bijection (§A.17.4). */
export interface TaskIdBijectionStore {
  getRemote(localTaskId: string): string | undefined;
  getLocal(remoteTaskId: string): string | undefined;
  bind(localTaskId: string, remoteTaskId: string): void;
  size(): number;
}

/**
 * In-memory bijection with an optional JSON file for durability across
 * adapter restarts. M2 keeps the file write synchronous and best-effort.
 */
export class MemoryTaskIdBijection implements TaskIdBijectionStore {
  private readonly localToRemote = new Map<string, string>();
  private readonly remoteToLocal = new Map<string, string>();
  private readonly filePath?: string;

  constructor(options?: { filePath?: string }) {
    this.filePath = options?.filePath;
    this.load();
  }

  getRemote(localTaskId: string): string | undefined {
    return this.localToRemote.get(localTaskId);
  }

  getLocal(remoteTaskId: string): string | undefined {
    return this.remoteToLocal.get(remoteTaskId);
  }

  bind(localTaskId: string, remoteTaskId: string): void {
    const existingRemote = this.localToRemote.get(localTaskId);
    if (existingRemote !== undefined && existingRemote !== remoteTaskId) {
      throw new Error(
        `task id bijection violated: ${localTaskId} already bound to ${existingRemote}`,
      );
    }
    const existingLocal = this.remoteToLocal.get(remoteTaskId);
    if (existingLocal !== undefined && existingLocal !== localTaskId) {
      throw new Error(
        `task id bijection violated: ${remoteTaskId} already bound to ${existingLocal}`,
      );
    }
    this.localToRemote.set(localTaskId, remoteTaskId);
    this.remoteToLocal.set(remoteTaskId, localTaskId);
    this.persist();
  }

  size(): number {
    return this.localToRemote.size;
  }

  entries(): Array<[string, string]> {
    return [...this.localToRemote.entries()];
  }

  private load(): void {
    if (!this.filePath) return;
    try {
      if (!existsSync(this.filePath)) return;
      const raw = JSON.parse(readFileSync(this.filePath, "utf8")) as Record<string, string>;
      for (const [local, remote] of Object.entries(raw)) {
        this.localToRemote.set(local, remote);
        this.remoteToLocal.set(remote, local);
      }
    } catch {
      // A corrupt store must not prevent adapter startup.
    }
  }

  private persist(): void {
    if (!this.filePath) return;
    try {
      writeFileSync(this.filePath, JSON.stringify(Object.fromEntries(this.localToRemote)), "utf8");
    } catch {
      // Best effort in M2.
    }
  }
}

/**
 * `MapOutboundTaskId` (§A.9.2 / §A.17.4). UUIDv7 local ids pass through
 * untouched; anything else is bound to a freshly minted UUIDv7 and reused on
 * every retry from the durable bijection.
 */
export function MapOutboundTaskId(
  localTaskId: string,
  store?: TaskIdBijectionStore,
  mint: () => string = uuidv7,
): string {
  if (isUuidV7(localTaskId)) return localTaskId;
  if (!store) return mint();
  const existing = store.getRemote(localTaskId);
  if (existing !== undefined) return existing;
  const minted = mint();
  store.bind(localTaskId, minted);
  return minted;
}

/** camelCase alias for call sites that prefer local naming conventions. */
export const mapOutboundTaskId = MapOutboundTaskId;


// ---------------------------------------------------------------------------
// Compatibility helpers used by adapter / idempotency / tests
// ---------------------------------------------------------------------------

export function polymeshStateToA2a(state: PolyMeshTaskState): A2ATaskState {
  return POLYMESH_TO_A2A_STATE[state];
}

export function a2aStateToPolymesh(state: A2ATaskState): PolyMeshTaskState {
  return A2A_TO_POLYMESH_STATE[state];
}

export function mayAdvanceState(prev: A2ATaskState | undefined, next: A2ATaskState): boolean {
  if (!prev) return true;
  if (prev === next) return true;
  if (isTerminalA2AState(prev)) return false;
  const rank: Record<A2ATaskState, number> = {
    submitted: 0,
    working: 1,
    completed: 9,
    failed: 9,
    canceled: 9,
  };
  return rank[next] >= rank[prev];
}

/** Adapter-friendly alias wrapping MemoryTaskIdBijection with get/set API. */
export class MemoryTaskIdMap {
  private readonly inner: MemoryTaskIdBijection;
  constructor(path?: string) {
    this.inner = new MemoryTaskIdBijection(path ? { filePath: path } : undefined);
  }
  get(local: string): string | undefined {
    return this.inner.getRemote(local);
  }
  set(local: string, remote: string): void {
    this.inner.bind(local, remote);
  }
  reverse(remote: string): string | undefined {
    return this.inner.getLocal(remote);
  }
  asStore(): TaskIdBijectionStore {
    return this.inner;
  }
}

export function buildTasksSendParams(input: {
  remoteTaskId: string;
  capability: string;
  payload: unknown;
  idempotency_key?: string;
  deadline?: string;
}): Record<string, unknown> {
  const skill = skillNameFromCapabilityName(input.capability);
  return {
    id: input.remoteTaskId,
    message: {
      role: "user",
      parts: [{ type: "data", data: input.payload }],
    },
    metadata: {
      skill,
      capability_id: input.capability,
      ...(input.idempotency_key ? { idempotency_key: input.idempotency_key } : {}),
      ...(input.deadline ? { deadline: input.deadline } : {}),
      skill_description: fidelityClause(input.capability),
    },
  };
}

export function fingerprintPayload(parts: {
  principal_id?: string;
  capability_id: string;
  input: unknown;
  task_id?: string;
  includeTaskId: boolean;
}): string {
  const body: Record<string, unknown> = {
    principal_id: parts.principal_id ?? "",
    capability_id: parts.capability_id,
    input: parts.input,
  };
  if (parts.includeTaskId && parts.task_id) body.task_id = parts.task_id;
  const canonical = JSON.stringify(sortKeysDeep(body));
  return createHash("sha256").update(canonical).digest("hex");
}

function sortKeysDeep(value: unknown): unknown {
  if (value === null || typeof value !== "object") return value;
  if (Array.isArray(value)) return value.map(sortKeysDeep);
  const obj = value as Record<string, unknown>;
  const out: Record<string, unknown> = {};
  for (const k of Object.keys(obj).sort()) out[k] = sortKeysDeep(obj[k]);
  return out;
}

File: src/polymesh/a2a/task_translator.py
"""Lifecycle translation between PolyMesh task states and A2A task states.

Covers §A.6 (state and progress mapping), §A.6.5 / §A.17.4 (shared task id
space with a durable bijection), and §A.10.6 (monotonic acceptance).
"""

from __future__ import annotations

import json
import os
import tempfile
from collections.abc import Callable, Mapping, Sequence
from pathlib import Path
from typing import Any

from ..protocol import uuidv7
from .card_mapper import skill_name_from_capability_name
from .errors import A2ADialectError, error_from_a2a_task_error
from .types import A2ASendParams, A2ATaskState, PolyMeshTaskState, TranslatedTaskEvent

A2A_TERMINAL_STATES: frozenset[str] = frozenset({"completed", "failed", "canceled"})

#: §A.6.3 inverse: A2A state -> PolyMesh state.
A2A_STATE_TO_POLYMESH: dict[str, PolyMeshTaskState] = {
    "submitted": "SUBMITTED",
    "working": "RUNNING",
    "completed": "SUCCEEDED",
    "failed": "FAILED",
    "canceled": "CANCELLED",
}

#: §A.6.3 forward: PolyMesh state -> A2A state.
POLYMESH_STATE_TO_A2A: dict[str, A2ATaskState] = {
    "SUBMITTED": "submitted",
    "ACCEPTED": "working",
    "QUEUED": "working",
    "RUNNING": "working",
    "WAITING": "working",
    "REJECTED": "failed",
    "SUCCEEDED": "completed",
    "FAILED": "failed",
    "CANCELLED": "canceled",
}

#: Monotonic rank used by §A.10.6; terminal states share the top rank.
A2A_STATE_RANK: dict[str, int] = {
    "submitted": 0,
    "working": 1,
    "completed": 2,
    "failed": 2,
    "canceled": 2,
}

_UUID_HYPHEN_POSITIONS = (8, 13, 18, 23)


def is_terminal_a2a_state(state: str | None) -> bool:
    return str(state) in A2A_TERMINAL_STATES


def is_uuidv7(value: Any) -> bool:
    """True when ``value`` is a canonical UUIDv7 string (version nibble 7)."""

    if not isinstance(value, str) or len(value) != 36:
        return False
    if any(value[index] != "-" for index in _UUID_HYPHEN_POSITIONS):
        return False
    hex_only = value.replace("-", "")
    if len(hex_only) != 32:
        return False
    try:
        int(hex_only, 16)
    except ValueError:
        return False
    return value[14] == "7" and value[19].lower() in {"8", "9", "a", "b"}


class TaskIdBijection:
    """Durable ``local_task_id <-> remote_task_id`` mapping (§A.17.4)."""

    def __init__(self, path: str | os.PathLike[str] | None = None) -> None:
        self._path = Path(path) if path is not None else None
        self._local_to_remote: dict[str, str] = {}
        self._remote_to_local: dict[str, str] = {}
        self._load()

    def _load(self) -> None:
        if self._path is None or not self._path.exists():
            return
        try:
            raw = json.loads(self._path.read_text("utf-8"))
        except (OSError, ValueError):
            return
        pairs = raw.get("pairs") if isinstance(raw, Mapping) else None
        if not isinstance(pairs, Mapping):
            return
        for local, remote in pairs.items():
            if isinstance(local, str) and isinstance(remote, str):
                self._local_to_remote[local] = remote
                self._remote_to_local[remote] = local

    def _persist(self) -> None:
        if self._path is None:
            return
        self._path.parent.mkdir(parents=True, exist_ok=True)
        payload = json.dumps({"pairs": self._local_to_remote}, separators=(",", ":"), sort_keys=True)
        handle = tempfile.NamedTemporaryFile(
            "w", encoding="utf-8", dir=str(self._path.parent), delete=False, prefix=".a2a-taskids-"
        )
        try:
            handle.write(payload)
            handle.flush()
            os.fsync(handle.fileno())
        finally:
            handle.close()
        os.replace(handle.name, self._path)

    def remote_for(self, local_task_id: str) -> str | None:
        return self._local_to_remote.get(local_task_id)

    def local_for(self, remote_task_id: str) -> str | None:
        return self._remote_to_local.get(remote_task_id)

    def bind(self, local_task_id: str, remote_task_id: str) -> str:
        existing = self._local_to_remote.get(local_task_id)
        if existing == remote_task_id:
            return existing
        self._local_to_remote[local_task_id] = remote_task_id
        self._remote_to_local[remote_task_id] = local_task_id
        self._persist()
        return remote_task_id

    def __len__(self) -> int:
        return len(self._local_to_remote)


def map_outbound_task_id(
    task_id: str,
    *,
    store: TaskIdBijection | None = None,
    mint: Callable[[], str] | None = None,
) -> str:
    """UUIDv7 passthrough, else mint a UUIDv7 and reuse it on every retry."""

    if not isinstance(task_id, str) or not 1 <= len(task_id) <= 128:
        raise A2ADialectError("MALFORMED", "task_id must be a 1-128 character string")
    if is_uuidv7(task_id):
        return task_id
    if store is not None:
        existing = store.remote_for(task_id)
        if existing is not None:
            return existing
    minted = (mint or uuidv7)()
    if store is not None:
        store.bind(task_id, minted)
    return minted


def build_send_params(
    *,
    remote_task_id: str,
    capability: str,
    payload: Any,
    idempotency_key: str | None = None,
    deadline: str | None = None,
    metadata: Mapping[str, Any] | None = None,
) -> A2ASendParams:
    """Build the ``tasks/send`` params for one outbound call (§A.9.2).

    The mesh capability id travels in metadata; the skill carries the stripped
    name so an A2A-native peer can route it (§A.5.1).
    """

    merged: dict[str, Any] = {"polymesh_capability_id": capability}
    if idempotency_key:
        merged["idempotency_key"] = idempotency_key
    if deadline:
        merged["deadline"] = deadline
    if metadata:
        merged.update(dict(metadata))
    return {
        "id": remote_task_id,
        "skill": skill_name_from_capability_name(capability),
        "message": {"role": "user", "parts": [{"type": "data", "data": payload}]},
        "metadata": merged,
    }


class MonotonicStateGate:
    """Reject regressive or post-terminal status updates (§A.10.6)."""

    def __init__(self) -> None:
        self._last_state: str | None = None
        self._last_rank: int = -1
        self._terminal = False

    @property
    def last_state(self) -> str | None:
        return self._last_state

    @property
    def terminal(self) -> bool:
        return self._terminal

    def accept(self, state: str) -> bool:
        rank = A2A_STATE_RANK.get(str(state))
        if rank is None:
            raise A2ADialectError("MALFORMED", f"unknown A2A task state {state!r}")
        if self._terminal:
            return False
        if rank < self._last_rank:
            return False
        self._last_state = str(state)
        self._last_rank = rank
        self._terminal = is_terminal_a2a_state(state)
        return True


def extract_artifact_result(task: Mapping[str, Any]) -> Any:
    """Pull the JSON result out of an A2A task's artifacts (§A.9.3)."""

    artifacts = task.get("artifacts")
    if not isinstance(artifacts, Sequence) or isinstance(artifacts, (str, bytes)):
        return None
    for artifact in artifacts:
        if not isinstance(artifact, Mapping):
            continue
        parts = artifact.get("parts")
        if not isinstance(parts, Sequence) or isinstance(parts, (str, bytes)):
            continue
        for part in parts:
            if not isinstance(part, Mapping):
                continue
            if "data" in part:
                return part["data"]
            text = part.get("text")
            if isinstance(text, str):
                try:
                    return json.loads(text)
                except ValueError:
                    return text
    return None


def _status_message(status: Mapping[str, Any]) -> str | None:
    message = status.get("message")
    if isinstance(message, str):
        return message
    if isinstance(message, Mapping):
        parts = message.get("parts")
        if isinstance(parts, Sequence) and not isinstance(parts, (str, bytes)):
            for part in parts:
                if isinstance(part, Mapping) and isinstance(part.get("text"), str):
                    return part["text"]
    return None


def translate_task_event(
    task: Mapping[str, Any],
    *,
    task_id: str,
    event_seq: int | None = None,
) -> TranslatedTaskEvent:
    """Project one remote A2A task snapshot onto a mesh lifecycle event."""

    if not isinstance(task, Mapping):
        raise A2ADialectError("MALFORMED", "A2A task snapshot must be an object")
    status = task.get("status")
    if not isinstance(status, Mapping):
        raise A2ADialectError("MALFORMED", "A2A task snapshot missing status object")
    state = status.get("state")
    if not isinstance(state, str) or state not in A2A_STATE_TO_POLYMESH:
        raise A2ADialectError("MALFORMED", f"unknown A2A task state {state!r}")

    mesh_state = A2A_STATE_TO_POLYMESH[state]
    hinted = task.get("metadata", {}).get("polymesh_state") if isinstance(task.get("metadata"), Mapping) else None
    if isinstance(hinted, str) and POLYMESH_STATE_TO_A2A.get(hinted) == state:
        mesh_state = hinted  # type: ignore[assignment]

    event: TranslatedTaskEvent = {
        "task_id": task_id,
        "state": mesh_state,
        "a2a_state": state,  # type: ignore[typeddict-item]
        "terminal": is_terminal_a2a_state(state),
    }
    remote_id = task.get("id")
    if isinstance(remote_id, str):
        event["remote_task_id"] = remote_id
    if event_seq is not None:
        event["event_seq"] = int(event_seq)

    progress = status.get("progress")
    if isinstance(progress, (int, float)) and not isinstance(progress, bool):
        event["progress"] = float(progress)
    message = _status_message(status)
    if message is not None:
        event["message"] = message

    if state == "completed":
        event["result"] = extract_artifact_result(task)
    elif state == "failed":
        error = status.get("error")
        mapped = error_from_a2a_task_error(error if isinstance(error, Mapping) else None, task_id=task_id)
        event["error"] = {
            "code": mapped.code,
            "message": str(mapped),
            "retryable": mapped.retryable,
            "json_rpc_code": mapped.json_rpc_code,
        }
    return event


__all__ = [
    "A2A_STATE_RANK",
    "A2A_STATE_TO_POLYMESH",
    "A2A_TERMINAL_STATES",
    "MonotonicStateGate",
    "POLYMESH_STATE_TO_A2A",
    "TaskIdBijection",
    "build_send_params",
    "extract_artifact_result",
    "is_terminal_a2a_state",
    "is_uuidv7",
    "map_outbound_task_id",
    "translate_task_event",
]
